### ISSUE TITLE
Sponsors.cash

### ISSUE THREAD

**Issue by baby636** [flipstarter/flipstarter.cash#32]
SPONSORS.CASH

https://cloudflare-ipfs.com/ipfs/Qmd4tk1a5KQtuondmdF1VD6CTEdWXNDJaHwiygwr4W6SzQ/#/welcome

Contributing to Bitcoin Core
============================

The Bitcoin Core project operates an open contributor model where anyone is
welcome to contribute towards development in the form of peer review, testing
and patches. This document explains the practical process and guidelines for
contributing.

First, in terms of structure, there is no particular concept of "Bitcoin Core
developers" in the sense of privileged people. Open source often naturally
revolves around a meritocracy where contributors earn trust from the developer
community over time. Nevertheless, some hierarchy is necessary for practical
purposes. As such, there are repository "maintainers" who are responsible for
merging pull requests, as well as a "lead maintainer" who is responsible for the
release cycle as well as overall merging, moderation and appointment of
maintainers.

Getting Started
---------------

New contributors are very welcome and needed.

Reviewing and testing is highly valued and the most effective way you can contribute
as a new contributor. It also will teach you much more about the code and
process than opening pull requests. Please refer to the [peer review](#peer-review)
section below.

Before you start contributing, familiarize yourself with the Bitcoin Core build
system and tests. Refer to the documentation in the repository on how to build
Bitcoin Core and how to run the unit tests, functional tests, and fuzz tests.

There are many open issues of varying difficulty waiting to be fixed.
If you're looking for somewhere to start contributing, check out the
[good first issue](https://github.com/bitcoin/bitcoin/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22)
list or changes that are
[up for grabs](https://github.com/bitcoin/bitcoin/issues?utf8=%E2%9C%93&q=label%3A%22Up+for+grabs%22).
Some of them might no longer be applicable. So if you are interested, but
unsure, you might want to leave a comment on the issue first.

You may also participate in the weekly
[Bitcoin Core PR Review Club](https://bitcoincore.reviews/) meeting.

### Good First Issue Label

The purpose of the `good first issue` label is to highlight which issues are
suitable for a new contributor without a deep understanding of the codebase.

However, good first issues can be solved by anyone. If they remain unsolved
for a longer time, a frequent contributor might address them.

You do not need to request permission to start working on an issue. However,
you are encouraged to leave a comment if you are planning to work on it. This
will help other contributors monitor which issues are actively being addressed
and is also an effective way to request assistance if and when you need it.

Communication Channels
----------------------

Most communication about Bitcoin Core development happens on IRC, in the
`#bitcoin-core-dev` channel on Libera Chat. The easiest way to participate on IRC is
with the web client, [web.libera.chat](https://web.libera.chat/#bitcoin-core-dev). Chat
history logs can be found
on [https://www.erisian.com.au/bitcoin-core-dev/](https://www.erisian.com.au/bitcoin-core-dev/)
and [https://gnusha.org/bitcoin-core-dev/](https://gnusha.org/bitcoin-core-dev/).

Discussion about codebase improvements happens in GitHub issues and pull
requests.

The developer
[mailing list](https://lists.linuxfoundation.org/mailman/listinfo/bitcoin-dev)
should be used to discuss complicated or controversial consensus or P2P protocol changes before working on
a patch set.

Contributor Workflow
--------------------

The codebase is maintained using the "contributor workflow" where everyone
without exception contributes patch proposals using "pull requests" (PRs). This
facilitates social contribution, easy testing and peer review.

To contribute a patch, the workflow is as follows:

  1. Fork repository ([only for the first time](https://docs.github.com/en/get-started/quickstart/fork-a-repo))
  1. Create topic branch
  1. Commit patches

For GUI-related issues or pull requests, the https://github.com/bitcoin-core/gui repository should be used.
For all other issues and pull requests, the https://github.com/bitcoin/bitcoin node repository should be used.

The master branch for all monotree repositories is identical.

As a rule of thumb, everything that only modifies `src/qt` is a GUI-only pull
request. However:

* For global refactoring or other transversal changes the node repository
  should be used.
* For GUI-related build system changes, the node repository should be used
  because the change needs review by the build systems reviewers.
* Changes in `src/interfaces` need to go to the node repository because they
  might affect other components like the wallet.

For large GUI changes that include build system and interface changes, it is
recommended to first open a pull request against the GUI repository. When there
is agreement to proceed with the changes, a pull request with the build system
and interfaces changes can be submitted to the node repository.

The project coding conventions in the [developer notes](doc/developer-notes.md)
must be followed.

### Committing Patches

In general, [commits should be atomic](https://en.wikipedia.org/wiki/Atomic_commit#Atomic_commit_convention)
and diffs should be easy to read. For this reason, do not mix any formatting
fixes or code moves with actual code changes.

Make sure each individual commit is hygienic: that it builds successfully on its
own without warnings, errors, regressions, or test failures.

Commit messages should be verbose by default consisting of a short subject line
(50 chars max), a blank line and detailed explanatory text as separate
paragraph(s), unless the title alone is self-explanatory (like "Corrected typo
in init.cpp") in which case a single title line is sufficient. Commit messages should be
helpful to people reading your code in the future, so explain the reasoning for
your decisions. Further explanation [here](https://chris.beams.io/posts/git-commit/).

If a particular commit references another issue, please add the reference. For
example: `refs #1234` or `fixes #4321`. Using the `fixes` or `closes` keywords
will cause the corresponding issue to be closed when the pull request is merged.

Commit messages should never contain any `@` mentions (usernames prefixed with "@").

Please refer to the [Git manual](https://git-scm.com/doc) for more information
about Git.

  - Push changes to your fork
  - Create pull request

### Creating the Pull Request

The title of the pull request should be prefixed by the component or area that
the pull request affects. Valid areas as:

  - `consensus` for changes to consensus critical code
  - `doc` for changes to the documentation
  - `qt` or `gui` for changes to bitcoin-qt
  - `log` for changes to log messages
  - `mining` for changes to the mining code
  - `net` or `p2p` for changes to the peer-to-peer network code
  - `refactor` for structural changes that do not change behavior
  - `rpc`, `rest` or `zmq` for changes to the RPC, REST or ZMQ APIs
  - `script` for changes to the scripts and tools
  - `test`, `qa` or `ci` for changes to the unit tests, QA tests or CI code
  - `util` or `lib` for changes to the utils or libraries
  - `wallet` for changes to the wallet code
  - `build` for changes to the GNU Autotools or reproducible builds

Examples:

    consensus: Add new opcode for BIP-XXXX OP_CHECKAWESOMESIG
    net: Automatically create onion service, listen on Tor
    qt: Add feed bump button
    log: Fix typo in log message

The body of the pull request should contain sufficient description of *what* the
patch does, and even more importantly, *why*, with justification and reasoning.
You should include references to any discussions (for example, other issues or
mailing list discussions).

The description for a new pull request should not contain any `@` mentions. The
PR description will be included in the commit message when the PR is merged and
any users mentioned in the description will be annoyingly notified each time a
fork of Bitcoin Core copies the merge. Instead, make any username mentions in a
subsequent comment to the PR.

### Translation changes

Note that translations should not be submitted as pull requests. Please see
[Translation Process](https://github.com/bitcoin/bitcoin/blob/master/doc/translation_process.md)
for more information on helping with translations.

### Work in Progress Changes and Requests for Comments

If a pull request is not to be considered for merging (yet), please
prefix the title with [WIP] or use [Tasks Lists](https://docs.github.com/en/github/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#task-lists)
in the body of the pull request to indicate tasks are pending.

### Address Feedback

At this stage, one should expect comments and review from other contributors. You
can add more commits to your pull request by committing them locally and pushing
to your fork until you have satisfied all feedback.

Note: Code review is a burdensome but important part of the development process, and as such, certain types of pull requests are rejected. In general, if the **improvements** do not warrant the **review effort** required, the PR has a high chance of being rejected. It is up to the PR author to convince the reviewers that the changes warrant the review effort, and if reviewers are "Concept NACK'ing" the PR, the author may need to present arguments and/or do research backing their suggested changes.

### Squashing Commits

If your pull request contains fixup commits (commits that change the same line of code repeatedly) or too fine-grained
commits, you may be asked to [squash](https://git-scm.com/docs/git-rebase#_interactive_mode) your commits
before it will be reviewed. The basic squashing workflow is shown below.

    git checkout your_branch_name
    git rebase -i HEAD~n
    # n is normally the number of commits in the pull request.
    # Set commits (except the one in the first line) from 'pick' to 'squash', save and quit.
    # On the next screen, edit/refine commit messages.
    # Save and quit.
    git push -f # (force push to GitHub)

Please update the resulting commit message, if needed. It should read as a
coherent message. In most cases, this means not just listing the interim
commits.

If you have problems with squashing or other git workflows, you can enable
"Allow edits from maintainers" in the right-hand sidebar of the GitHub web
interface and ask for help in the pull request.

Please refrain from creating several pull requests for the same change.
Use the pull request that is already open (or was created earlier) to amend
changes. This preserves the discussion and review that happened earlier for
the respective change set.

The length of time required for peer review is unpredictable and will vary from
pull request to pull request.

### Rebasing Changes

When a pull request conflicts with the target branch, you may be asked to rebase it on top of the current target branch.
The `git rebase` command will take care of rebuilding your commits on top of the new base.

This project aims to have a clean git history, where code changes are only made in non-merge commits. This simplifies
auditability because merge commits can be assumed to not contain arbitrary code changes. Merge commits should be signed,
and the resulting git tree hash must be deterministic and reproducible. The script in
[/contrib/verify-commits](/contrib/verify-commits) checks that.

After a rebase, reviewers are encouraged to sign off on the force push. This should be relatively straightforward with
the `git range-diff` tool explained in the [productivity
notes](/doc/productivity.md#diff-the-diffs-with-git-range-diff). To avoid needless review churn, maintainers will
generally merge pull requests that received the most review attention first.

Pull Request Philosophy
-----------------------

Patchsets should always be focused. For example, a pull request could add a
feature, fix a bug, or refactor code; but not a mixture. Please also avoid super
pull requests which attempt to do too much, are overly large, or overly complex
as this makes review difficult.

### Features

When adding a new feature, thought must be given to the long term technical debt
and maintenance that feature may require after inclusion. Before proposing a new
feature that will require maintenance, please consider if you are willing to
maintain it (including bug fixing). If features get orphaned with no maintainer
in the future, they may be removed by the Repository Maintainer.

### Refactoring

Refactoring is a necessary part of any software project's evolution. The
following guidelines cover refactoring pull requests for the project.

There are three categories of refactoring: code-only moves, code style fixes, and
code refactoring. In general, refactoring pull requests should not mix these
three kinds of activities in order to make refactoring pull requests easy to
review and uncontroversial. In all cases, refactoring PRs must not change the
behaviour of code within the pull request (bugs must be preserved as is).

Project maintainers aim for a quick turnaround on refactoring pull requests, so
where possible keep them short, uncomplex and easy to verify.

Pull requests that refactor the code should not be made by new contributors. It
requires a certain level of experience to know where the code belongs to and to
understand the full ramification (including rebase effort of open pull requests).

Trivial pull requests or pull requests that refactor the code with no clear
benefits may be immediately closed by the maintainers to reduce unnecessary
workload on reviewing.

"Decision Making" Process
-------------------------

The following applies to code changes to the Bitcoin Core project (and related
projects such as libsecp256k1), and is not to be confused with overall Bitcoin
Network Protocol consensus changes.

Whether a pull request is merged into Bitcoin Core rests with the project merge
maintainers and ultimately the project lead.

Maintainers will take into consideration if a patch is in line with the general
principles of the project; meets the minimum standards for inclusion; and will
judge the general consensus of contributors.

In general, all pull requests must:

  - Have a clear use case, fix a demonstrable bug or serve the greater good of
    the project (for example refactoring for modularisation);
  - Be well peer-reviewed;
  - Have unit tests, functional tests, and fuzz tests, where appropriate;
  - Follow code style guidelines ([C++](doc/developer-notes.md), [functional tests](test/functional/README.md));
  - Not break the existing test suite;
  - Where bugs are fixed, where possible, there should be unit tests
    demonstrating the bug and also proving the fix. This helps prevent regression.
  - Change relevant comments and documentation when behaviour of code changes.

Patches that change Bitcoin consensus rules are considerably more involved than
normal because they affect the entire ecosystem and so must be preceded by
extensive mailing list discussions and have a numbered BIP. While each case will
be different, one should be prepared to expend more time and effort than for
other kinds of patches because of increased peer review and consensus building
requirements.

### Peer Review

Anyone may participate in peer review which is expressed by comments in the pull
request. Typically reviewers will review the code for obvious errors, as well as
test out the patch set and opine on the technical merits of the patch. Project
maintainers take into account the peer review when determining if there is
consensus to merge a pull request (remember that discussions may have been
spread out over GitHub, mailing list and IRC discussions).

#### Conceptual Review

A review can be a conceptual review, where the reviewer leaves a comment
 * `Concept (N)ACK`, meaning "I do (not) agree with the general goal of this pull
   request",
 * `Approach (N)ACK`, meaning `Concept ACK`, but "I do (not) agree with the
   approach of this change".

A `NACK` needs to include a rationale why the change is not worthwhile.
NACKs without accompanying reasoning may be disregarded.

#### Code Review

After conceptual agreement on the change, code review can be provided. A review
begins with `ACK BRANCH_COMMIT`, where `BRANCH_COMMIT` is the top of the PR
branch, followed by a description of how the reviewer did the review. The
following language is used within pull request comments:

  - "I have tested the code", involving change-specific manual testing in
    addition to running the unit, functional, or fuzz tests, and in case it is
    not obvious how the manual testing was done, it should be described;
  - "I have not tested the code, but I have reviewed it and it looks
    OK, I agree it can be merged";
  - A "nit" refers to a trivial, often non-blocking issue.

Project maintainers reserve the right to weigh the opinions of peer reviewers
using common sense judgement and may also weigh based on merit. Reviewers that
have demonstrated a deeper commitment and understanding of the project over time
or who have clear domain expertise may naturally have more weight, as one would
expect in all walks of life.

Where a patch set affects consensus-critical code, the bar will be much
higher in terms of discussion and peer review requirements, keeping in mind that
mistakes could be very costly to the wider community. This includes refactoring
of consensus-critical code.

Where a patch set proposes to change the Bitcoin consensus, it must have been
discussed extensively on the mailing list and IRC, be accompanied by a widely
discussed BIP and have a generally widely perceived technical consensus of being
a worthwhile change based on the judgement of the maintainers.

### Finding Reviewers

As most reviewers are themselves developers with their own projects, the review
process can be quite lengthy, and some amount of patience is required. If you find
that you've been waiting for a pull request to be given attention for several
months, there may be a number of reasons for this, some of which you can do something
about:

  - It may be because of a feature freeze due to an upcoming release. During this time,
    only bug fixes are taken into consideration. If your pull request is a new feature,
    it will not be prioritized until after the release. Wait for the release.
  - It may be because the changes you are suggesting do not appeal to people. Rather than
    nits and critique, which require effort and means they care enough to spend time on your
    contribution, thundering silence is a good sign of widespread (mild) dislike of a given change
    (because people don't assume *others* won't actually like the proposal). Don't take
    that personally, though! Instead, take another critical look at what you are suggesting
    and see if it: changes too much, is too broad, doesn't adhere to the
    [developer notes](doc/developer-notes.md), is dangerous or insecure, is messily written, etc.
    Identify and address any of the issues you find. Then ask e.g. on IRC if someone could give
    their opinion on the concept itself.
  - It may be because your code is too complex for all but a few people, and those people
    may not have realized your pull request even exists. A great way to find people who
    are qualified and care about the code you are touching is the
    [Git Blame feature](https://docs.github.com/en/github/managing-files-in-a-repository/managing-files-on-github/tracking-changes-in-a-file). Simply
    look up who last modified the code you are changing and see if you can find
    them and give them a nudge. Don't be incessant about the nudging, though.
  - Finally, if all else fails, ask on IRC or elsewhere for someone to give your pull request
    a look. If you think you've been waiting for an unreasonably long time (say,
    more than a month) for no particular reason (a few lines changed, etc.),
    this is totally fine. Try to return the favor when someone else is asking
    for feedback on their code, and the universe balances out.
  - Remember that the best thing you can do while waiting is give review to others!

Backporting
-----------

Security and bug fixes can be backported from `master` to release
branches.
If the backport is non-trivial, it may be appropriate to open an
additional PR to backport the change, but only after the original PR
has been merged.
Otherwise, backports will be done in batches and
the maintainers will use the proper `Needs backport (...)` labels
when needed (the original author does not need to worry about it).

A backport should contain the following metadata in the commit body:

```
Github-Pull: #<PR number>
Rebased-From: <commit hash of the original commit>
```

Have a look at [an example backport PR](
https://github.com/bitcoin/bitcoin/pull/16189).

Also see the [backport.py script](
https://github.com/bitcoin-core/bitcoin-maintainer-tools#backport).

Release Policy
--------------

The project leader is the release manager for each Bitcoin Core release.

Copyright
---------

By contributing to this repository, you agree to license your work under the
MIT license unless specified otherwise in `contrib/debian/copyright` or at
the top of the file itself. Any work contributed where you are not the original
author must contain its license header with the original author(s) and source.

UNSTOPPABLE
MENU
How to become a BCH Full Stack Developer
This is a guide to becoming a Bitcoin Cash (BCH) full stack developer. It's a deep rabbit hole and a long journey to mastering the full stack of BCH software, but the reward is making modern web and mobile apps that speak directly to the blockchain. But first, lets start with the most basic question: What is a BCH full stack developer? What does that term even mean?

What is a BCH full stack developer?
Let's start with a couple definitions:

Developer: A person who writes software. Because the world of software is so large, most developers will distinguish themselves by picking a specific programming language like JavaScript, Python, Java, or C#.

Full Stack: Developers will typically specialize in front end development: phone apps, web apps, or other things with a user interface (UI). Or back end development: web servers, desktop applications, anything not involving a user interface. A full stack developer is one who is proficient at both.
Let's start with the end goal and work backwards from there: A BCH full stack developer, with the skills that businesses are willing to pay good money for, can build a wallet application or some other application that talks directly to the BCH blockchain. It can reliably receive and send transactions. They are able to run and maintain the infrastructure required to support that application.

This guide will give you a road map to acquiring the skills necessary to become a BCH full stack developer. The emphasis is on JavaScript, but the infrastructure applies to any programming language.

Dipping your toes in the water
Before digging into the nitty-gritty back end infrastructure, it's often encouraging to 'skip ahead' to the end and show yourself why you're about to work so hard to learn all that infrastructure. Here are some examples of popular front end applications. Each of these apps use back end infrastructure that this article will show you how to set up.

Badger Wallet is really two wallet apps. One is for smart phones and the other is for web browsers. But they function the same. This is one of the leading wallets for working with SLP tokens.

Memo is a clone of Twitter, but uses the BCH blockchain. Because its based on the blockchain, there can be no deplatforming or censorship of any kind. Memo also includes a decentralized exchange for selling SLP tokens.

read.cash is a blogging platform similar to Medium. The site implements a BCH wallet and 'micro-tipping'. You can tip an auther as little at $0.05 to show your appriciation for their work. Comments are upvoted based on these micro tips, which leads to really healthy discussions about the content.

Satoshi Dice is a very simple gambling game that lets you bet with BCH. Because is uses the BCH blockchain, all bets and payouts are publicly transparent and the game is provably fair.
This wide range of applications all use the same common back end infrastructure.

The Full BCH Stack
The diagram below illustrates the stack of software required to build applications like the ones above. Those applications all fit into the top-most layer of the stack.

The BCH Full Stack
The BCH Full Stack
Here is a description of each layer, starting at the bottom and working up to the top:

Full Node - A full node is the most fundamental piece of software in Bitcoin. This is a 'non-pruned' or 'archival' node that contains a full copy of the blockchain.

Indexers - Indexers are like small search engines. They communicate with the full node and crawl the blockchain data to create databases that contain additional metadata that is not directly accessible from the full node. This metadata includes things like SLP tokens, and posts on Memo.cash.

REST API - A REST API web server provides an interface for all modern programming languages to communicate indirectly with the indexers and full node. This layer provides a common interface for modern software (like web and mobile apps) to speak to the blockchain.

JavaScript Library - A JavaScript library like bitbox-sdk or bch-js provides an easy way to communicate with the REST API layer. It also contains utility software for common things like crafting custom transactions, creating wallets, and signing messages.

Applications - Finally, the application layer is where the magic happens. This is the business logic that makes one app different from another. This is the user interface that end-users get to interact with. This is the layer where front end developers play. The examples listed in the last section all fit into this layer.
This entire software stack can be run on a modern computer with 16-32 GB of RAM, 1-2 TB of hard drive space, and a decent internet connection.

I'll explore each layer in more detail in the sections below. I'll also link to Docker containers for running each layer of the stack.

Skipping the back end
To help front end developers get started quickly, and to remove the burden of hosting your own back end infrastructure, there are a growing list of cloud service providers. This provides convenience at a price.

Note: Trusting a service provider to handle the back end infrastructure for you is convenient, but if their servers go down, your app goes down with them. And unless you run your own fall-back infrastructure, you have no recourse but to simply wait until they get back up and running. This is known as platform risk. The only way to eliminate platform risk is to run your own infrastructure.

Here are a few of the cloud-based REST APIs for interacting with the BCH network:

rest.bitcoin.com is probably the most popular REST API for interacting with the BCH blockchain. bitbox-sdk is a JavaScript library that provides easy endpoints for working with rest.bitcoin.com and provides additional important tools for working with Bitcoin Cash. Documentation for both can be found at developer.bitcoin.com.

FullStack.cash is a pay-to-play fork of rest.bitcoin.com that is accessed through api.bchjs.cash. bch-js is a fork of the bitbox-sdk JavaScript library that is designed to work with it.

bloq.cloud and Block Cypher are a pay-to-play REST APIs with similar access as the above options, but they have very different structures to their REST API. While it's easy to switch between Bitcoin.com tools and FullStack.cash tools (because they are forks of the same code base), there is much more work involved with switching to these other platforms.

There are very few free or paid REST API services for working with BCH. The industry needs more, and you could profit by providing these resources. This article outlines the steps you can take to run your own cloud infrastructure, to earn BCH as illustrated in this video:

Running a Full Node
The full node is the most fundamental piece of software. To run a full node, you'll need enough hard drive space to store the blockchain. This is currently under 200GB, but it's always growing. You'll also need at least 2 GB of RAM. Syncing a full node can take anywhere from 2 days to 2 weeks, depending on your hardware and internet connection.

There are several full node implementations. Here are a few:

ABC has the majority of market share. It's the implementation that started the BCH fork, and is the most widely used by exchanges and miners.

BU is the second most popular Bitcoin Cash implementation.

BCHD is a rapidly growing, developer-friendly implementation written in Go.

Flowee the Hub is a one of the newest implementations on the scene.

Bitcoin Verde is another full node implementation.
This article only focuses on the ABC implementation of Bitcoin Cash. I maintain a Docker container for running this full node. Installation instructions can be found in the GitHub repository for docker-abc.

Indexers
As mentioned at the beginning of this article, indexers are like little search engines that crawl the blockchain and stitch together metadata that is not directly available from the full node. The two most important features that an indexer provides for wallet apps and other modern applications is:

BCH balance lookup by address
UTXO lookup by address
A couple widely used, open source indexers are:

Blockbook
Bitcore Node
Those links will take you to Docker containers that I maintain for both indexers.

A few up-and-coming indexers include:

BCHD has indexing capabilities via its startup options and its gRPC interface.
Flowee Indexer is a new, experimental indexer.
SLP tokens
The Simple Ledger Protocol uses an indexer to validate token transactions. SLP tokens are one of the newest, most popular protocols that ride on top of Bitcoin Cash. Each SLP token is a 'colored' UTXO. Token transactions are tracked using OP_RETURN data in transactions. The indexer follows the trail of transactions to track token balances.

My Dockerized SLPDB includes SLPDB, a Mongo database, and the API front end, slpserve. It's the complete package for interfacing to the SLP token world.

REST API
The REST API is the communication interface that lets modern web and mobile apps talk to the blockchain. It's a lightweight webserver that wraps the indexers and full node in a REST API. This allows high level applications to talk to low-level interfaces.

Some of the REST API services that exist:

rest.bitcoin.com is an extension of developer.bitcoin.com.

FullStack.cash runs bch-api, a fork of rest.bitcoin.com.

bch-api has been adapted to work with this front end and this back end authorization server. It generates JWT tokens for charging for access to your API, or for securing your API, so that when exposing your API to the internet, random public people can't use it.
I have not tried out these other REST APIs, but they look interesting:

bloq.cloud
Block Cypher
Blockchain.com
JavaScript Libraries
bitbox-sdk is one of the most popular JavaScript libraries for working with Bitcoin Cash. You can look at the documentation and browse all the features. Not only does it provide an easy interface to the REST API, it includes a complete toolbox of commonly needed methods when working with Bitcoin Cash. Browse simple example apps to learn how to code with it quickly.

slp-sdk is a superset of bitbox-sdk. It contains all the functionality of bitbox-sdk, plus the ablility to work with SLP tokens. It also has examples and documentation.

bch-js is a fork of bitbox-sdk, which works with the FullStack.cash API. It also has a collection of example apps and documentation. It has a few additional features compared to the libraries above:

Interface with the OpenBazaar Indexer
Works with SLP tokens natively (no external libraries).
Includes endpoints for Blockbook and Bitcore indexers.
Support for BCHD comming soon!
Applications
Finally! We're at the application layer! Yay!

While bitbox-sdk, slp-sdk, and bch-js provide important functionality, most application developers are disappointed in how much bitcoin-specific knowledge is required to interact with them. That's why I created slp-cli-wallet. It's a command-line interface (CLI) app, but it's also an npm library that can easily be included into any other application. It provides high level functions like create-wallet, get-address, and send-bch. It reduces the amount of bitcoin-specific knowledge required to create applications.

Speaking of required knowledge, all developers in this space should read and frequently refer to Mastering Bitcoin Cash. It'll be hard to go far in this space without familiarizing yourself with basic Bitcoin concepts. The book is free and one of the best resources.

Also, there is Badger SDK. This provides React components and other code examples for interacting with Badger Wallet. If you want to replace Paypal buttons with BCH buttons, this is the fastest way to do it.

Minimum Viable Infrastructure
There is a concept in the startup world of a minimum viable product or MVP. There are so many options of full nodes, indexers, and libraries that its easy to waste a lot of time on a less-than-optimal solution. I've been working full-time on BCH infrastructure for two years. Here is the advice I have to give to aspiring developers who want to set up their own minimum viable infrastructure, it's based on the downloadable databases here.

Use this ABC full node Docker container. It's the most popular full node implementation, and it's hard to go wrong with that choice.

Use Blockbook for your indexer. It has the smallest database and has shown to be the most reliable. This Docker container allows you to run your own BCH version locally.

If you plan to dip your feet into the world of SLP tokens, then you'll want to run this Docker container. It contains everything needed to interact with them.

bch-api is a self-hosted REST API that will communicate with all the above software. It's a node.js JavaScript web server using the express.js framework, and is a fork of rest.bitcoin.com.

bch-js is a JavaScript library that is intended to be paired with bch-api. It's also a fork of bitbox-sdk.

slp-cli-wallet is an application-level JavaScript library that is intended to be paired with bch-js. It provides very high-level calls for applications to interact with the BCH blockchain.

It will take weeks to sync the full node and indexers on a home computer. To shorten this time, you can download pre-synced databases that contain the last eleven years of blockchain data. Those files are intended to work with the above Docker containers. It should reduce the time syncing to a couple hours.

A final tip: Avoid testnet. It's traditional to develop apps using the testnet so that you don't waste real money. But the reality is that testnet is often too flaky to use, for a number of reasons. At the same time, it's possible to send 1000 sat (0.00001 BCH, a fraction of a penny) transactions on mainnet, with a mining fee of 1 sat/byte, while maintaining a high confidence that the transaction will be included in the next block. That makes a compelling argument to use mainnet for app testing instead of testnet. Most developers I know, including myself, take this approach.

Getting Support
The Bitcoin Cash space is a rapidly moving target. The core concepts in this article won't change much, but the tools, implementations, and libraries are in a constant state of flux. A big part of being a successful full stack developer is injecting yourself into the community, so that you can ride the flow of ever-changing information. Most of the community discourse takes place on Telegram:

bch-js-toolkit is a public channel for asking questions and getting support for bch-js, bch-api, and the FullStack.cash cloud service.

Permissionless Software is a channel for discussing PSF inititives like FullStack.cash, its business plan, the PSF token, and other software sponsored by the Permissionless Software Foundation

Simpleledger is the place to explore and ask questions around SLP tokens.

Fountainhead is a public channel run by a group of experienced BCH developers.
The above channels are a great place to get introduced to the Bitcoin Cash developer community. There are many other private channels (requiring an invite) where the real development discussion happens. Asking sincere questions and showing off your work in the channels above are the best way to find access to those channels.

A lot of discussion and sharing also happens on Twitter. You can follow me @christroutner. I follow and interact with many BCH full stack developers on Twitter, and you can too.

I can't wait to see what you build!

Synidicated over IPFS
Updates announced on this memo.cash account
https://read.cash/@krise636/bitcoin-core-f77c629c

memo-push
A simple command-line tool to publish IPFS hashes to the BCH blockchain using the memo.cash protocol.

oclif Version Downloads/week License Greenkeeper badge

Installation
Install the npm package globally, in order to use it as a command line tool:
sudo npm install -g memo-push

Create an account on memo.cash and add the private key (in WIF compressed format) to your environment variables:
export WIF=L4b9LdXgVUHYXWmSV1VFXhUzZjXhtHg6mHcx4j8XfYG9K6fReMid

Note: Any BCH address WIF can be used, but if WIF matches a memo.cash account, people will easily be able to see your updates by following your memo.cash profile.

Usage
memo-push -h - display help menu
memo-push -v - display version
memo-push -p <hash> - Publish an IPFS hash to the BCH blockchain.
Runkit
search your notebooks
Your notebooks
Untitled - /baby636/60e171d14ac9db001d4de764
Last edited 2 months ago
1
​
Untitled - /baby636/60e171e2837e54001adf0645
Last edited 3 months ago
1
sudo add-apt-repository ppa:certbot/certbot
2
sudo apt-get update
3
sudo apt-get install python-certbot-nginx
4
sudo certbot --nginx -d yourdomain.com
5
​
6
​
7
 docker build -t flipstarter .
8
docker volume create flipstarter
9
docker run -v flipstarter:/app/static/campaigns -p 3000:3000 flipstarter
10
​
11
 npm install
12
npm start
13
​
14
# or on windows:
15
npm run start-no-debug
16
​
17
 # download the latest version of flipstarter
18
docker pull flipstarter/flipstarter
19
​
20
# create a volume to store the database
21
docker volume create flipstarter
22
​
23
# run the container on ports 443 and 80
24
docker run -d --restart always --name flipstarter -v flipstarter:/app/static/campaigns -p 443:3000 -p 80:3000 flipstarter/flipstarter
25
​
26
​
27
 https://ipfs.io/ipns/k51qzi5uqu5dgdocf63983yb9eigz3nqsu892vph2dq9hsxm5bhj8c45o5ggdb/build.svg)](https://github.com/SahidMiller/ipfs-flipstarter/actions)
28
​
29
​
30
#!/usr/bin/env nodevar debug = require('debug')('my-application');var app = require('../app'); app.set('port', process.env.PORT || 3002); var server = app.listen(app.get('port'), function() { debug('Express server listening on port ' + server.address().port);     if (app.runOnStartup) {     app.runOnStartup(); }});
31
​
32
​
33
https://api.netlify.com/api/v1/badges/1f511982-8cee-47de-9246-6c07a792072e/
34
​
35
​
36
https://darkedge.app/
37
​
38
https://bitcoinapps.info
39
​
40
https://crowdin.com/project/bchnod
41
​
42
https://badges.crowdin.net/bchnode
43
​
44
https://member.cash/t/flipstarter
45
​
46
http://www.w3.org/2005/Atom
47
​
48
https://secure.gravatar.com/avatar/0309579a736fe1632714b306f8d1702a
49
​
50
https://cloudflare-ipfs.com/ipfs/Qmd4tk1a5KQtuondmdF1VD6CTEdWXNDJaHwiygwr4W6SzQ/#/welcome
51
​
52
https://auth.pinata.cloud
53
​
54
http://momentjs.com/
55
​
56
https://flipstarter.too.krise636.cash/
57
​
58
https://cloudflare-ipfs.com/ipfs/Qmd4tk1a5KQtuondmdF1VD6CTEdWXNDJaHwiygwr4W6SzQ/#/welcome
59
​
60
Client ID
61
0I1pqXeyfDVy4qyNnIhpTg4G3IoUnIdY3iANsx6k
62
​
63
  flipstarter.causes.cash-baby636
64
 
65
 http://
66
Onlycoin.app
67
​
68
​
JSON Endpoint Example 2 - /baby636/60e17149f3976a001bdee9a0
Last edited 3 months ago
Send a request like ?url=http://google.com, get back the title of the page.
RunKit + npm: @chris.troutner/bch-js - /baby636/60e1710ebe4786001a4292ab
Last edited 3 months ago - from: https://npm.runkit.com/%40chris.troutner%2Fbch-js
1
var bchJs = require("@chris.troutner/bch-js")
RunKit + npm: @psf/bch-js - /baby636/60cef8aee2702b001abc6b68
Last edited 3 months ago - from: https://npm.runkit.com/%40psf%2Fbch-js
1
var bchJs = require("@psf/bch-js")
ES6 and ES7 Support - /baby636/60cae2205abff4001a786abd
Last edited 4 months ago
ECMAScript 6 (also known as ES2015) Support:
​
Arrow Functions: https://tonicdev.com/tonic/es6-arrow-functions
Destructuring: https://tonicdev.com/tonic/es6-destructuring
Enhanced object literals: https://tonicdev.com/tonic/es6-enhanced-object-literals
Default, Rest, and Spread: https://tonicdev.com/tonic/es6-default-rest-and-spread
Map, Set, WeakMap, WeakSet: https://tonicdev.com/tonic/es6-maps-weakmaps-sets-and-weaksets
Promises: (description coming soon)
Binary and octal literals: https://tonicdev.com/tonic/es6-binary-and-octal-literals
Symbols: https://tonicdev.com/tonic/es6-symbols
Generators and Iterators: https://tonicdev.com/tonic/es6-iterators-and-generators
Template Strings: https://tonicdev.com/tonic/es6-template-strings
​
ECMAScript 7 Support:
​
Asynchronous Functions: https://tonicdev.com/tonic/es7-async-functions
​
Extra Support:
​
JSX and React: https://tonicdev.com/tonic/jsx-in-tonic
Untitled - /baby636/60cae07f1e5c92001a89b28b
Last edited 4 months ago
1
var GitHubApi = require("@octokit/rest");
2
var github = new GitHubApi({
3
    version: "3.0.0"
4
});
5
​
6
github.authenticate({
7
    type: "oauth",
8
    token: process.env.GITHUB_ACCESS_TOKEN
9
});
10
​
11
await github.repos.getAll({});require("bit-python")
Untitled - /baby636/60cadeb2800c59001a1de8e5
Last edited 4 months ago
1
​
Untitled - /baby636/60cade64f39274001b2f42ec
Last edited 4 months ago
1
return
RunKit + npm: bcoin - /baby636/60cade539400a4001ac53d55
Last edited 4 months ago
1
var bcoin = require("bcoin")
«
1
2
3
»

ECMAScript 6 (also known as ES2015) Support:

Arrow Functions: https://tonicdev.com/tonic/es6-arrow-functions
Destructuring: https://tonicdev.com/tonic/es6-destructuring
Enhanced object literals: https://tonicdev.com/tonic/es6-enhanced-object-literals
Default, Rest, and Spread: https://tonicdev.com/tonic/es6-default-rest-and-spread
Map, Set, WeakMap, WeakSet: https://tonicdev.com/tonic/es6-maps-weakmaps-sets-and-weaksets
Promises: (description coming soon)
Binary and octal literals: https://tonicdev.com/tonic/es6-binary-and-octal-literals
Symbols: https://tonicdev.com/tonic/es6-symbols
Generators and Iterators: https://tonicdev.com/tonic/es6-iterators-and-generators
Template Strings: https://tonicdev.com/tonic/es6-template-strings

ECMAScript 7 Support:

Asynchronous Functions: https://tonicdev.com/tonic/es7-async-functions

Extra Support:

JSX and React: https://tonicdev.com/tonic/jsx-in-tonic

FundMe.cash Whitepaper
Abstract:
Be a part of Fundme.Cash improvements to help users get funded faster and make this
fundraising platform more transparent, flexible for donors, and profitable for investors.
We are no longer taking fees from users, therefore this platform will be powered by our SLP
token. (FMT - FundMeToken)
All campaign participants will get FMT**
What is fundme.cash?
This is a BCH fundraising platform. Users can create a campaign in under 5 minutes.
Currently we do not have a for-profit-model. We intend to change that for investors and the
fundme.cash team.
Proposal:
Goal
Update fundme.cash with new features:
● Create and distribute FMT
● Map and build FMT rewards/referral system
● Map and build BCH advertisement bidding system
● Remove fees from platform
● Code new update system
● Integrate telegram notifications
● Create campaign wallet on client side
● Create BCH to FMT swap
● Integrate dividend payment for profit airdrops

https://t.me/flipfund/632
Read.Cash/pre-flip

Read.Cash/pre-flip
https://read.cash/@krise636/bitcoin-core-f77c629c

https://www.chronicheather.eu.org/
https://chainchecker.certifytheweb.com/#:~:text=7.0%20and%20lower.-,flipstarter.cash,OK,-An%20important%20bug

sudo add-apt-repository ppa:certbot/certbot
sudo apt-get update
sudo apt-get install python-certbot-nginx
sudo certbot --nginx -d yourdomain.com

 docker build -t flipstarter .
docker volume create flipstarter
docker run -v flipstarter:/app/static/campaigns -p 3000:3000 flipstarter

 npm install
npm start

# or on windows:
npm run start-no-debug

 # download the latest version of flipstarter
docker pull flipstarter/flipstarter

# create a volume to store the database
docker volume create flipstarter

# run the container on ports 443 and 80
docker run -d --restart always --name flipstarter -v flipstarter:/app/static/campaigns -p 443:3000 -p 80:3000 flipstarter/flipstarter

 https://ipfs.io/ipns/k51qzi5uqu5dgdocf63983yb9eigz3nqsu892vph2dq9hsxm5bhj8c45o5ggdb/build.svg)](https://github.com/SahidMiller/ipfs-flipstarter/actions)

#!/usr/bin/env nodevar debug = require('debug')('my-application');var app = require('../app'); app.set('port', process.env.PORT || 3002); var server = app.listen(app.get('port'), function() {	debug('Express server listening on port ' + server.address().port); 	if (app.runOnStartup) {		app.runOnStartup();	}});

https://api.netlify.com/api/v1/badges/1f511982-8cee-47de-9246-6c07a792072e/

https://darkedge.app/

https://bitcoinapps.info

https://crowdin.com/project/bchnod

https://badges.crowdin.net/bchnode

https://member.cash/t/flipstarter

http://www.w3.org/2005/Atom

https://secure.gravatar.com/avatar/0309579a736fe1632714b306f8d1702a

https://cloudflare-ipfs.com/ipfs/Qmd4tk1a5KQtuondmdF1VD6CTEdWXNDJaHwiygwr4W6SzQ/#/welcome

https://auth.pinata.cloud

http://momentjs.com/

https://flipstarter.too.krise636.cash/

https://cloudflare-ipfs.com/ipfs/Qmd4tk1a5KQtuondmdF1VD6CTEdWXNDJaHwiygwr4W6SzQ/#/welcome

flipstarter.causes.cash-baby636
 
 http://Onlycoin.app
